### PR TITLE
Option to change page gap in continuous mode

### PR DESCRIPTION
**Pull request by mustafa-001**
for request see https://github.com/koreader/koreader/issues/5656

Adds an option to select page gap in continuous mode. Option only shows up on documents that have numbered pages (.pdf, .cbz) because i think there is not much use to this option when reading arbitrarily paged documents. If seen necessary same logic can be duplicated for .epub files.

koptoptions.lua handles user selection and saving that selection
ReaderView:readSettings() handles reading previously selected option when opening book
ReaderVİew:onPageGapUpdate() renders changes when selection made

Page gap is same in both landscape and portrait modes, I think adding different settings for both is little unnecessary. 
The options for gap is 0, 8, 16 and 32, default one before was 8.

commit log
> 
> page_gap setting on opening a book
> if it does not exist defaults to 8
> -adds page_gap selector to koptoptions.lua,
> user can select 0, 8, 16, 32 via footer settings
> -adds :onPageGapUpdate event handler to ReaderView

page-1
page-gap2

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/5705)

### CODE REVIEW

**Comment by mustafa-001** [comment]
https://github.com/koreader/koreader/blob/bfe9cd8527db340ad73bcc0f8466df205081a133/frontend/ui/data/koptoptions.lua#L8-L11

Changes on this settings forces screen wide refresh, and ReaderView:painttto() draws everything again. Also I didn't quite figure out what to do other than full screen refresh thats already queued :). Any suggestions on proper way of doing this are welcomed.

**Comment by NiLuJe** [comment]
Yeah, IIRC, koptsettings can force repaints without any specific events because... reasons? :D.

**Comment by Frenzie** [comment]
It probably mostly makes sense to just redraw everything after changing a setting… a bit wasteful depending on what it is perhaps but I doubt it's bothersome.

So if that's not an issue, it seems fine to me.

File: frontend/apps/reader/modules/readerview.lua [hunk 1]
```diff
@@ -782,6 +782,8 @@ function ReaderView:onReadSettings(config)
     self.page_scroll = page_scroll == 1 and true or false
     self.highlight.saved = config:readSetting("highlight") or {}
     self.page_overlap_style = config:readSetting("page_overlap_style") or G_reader_settings:readSetting("page_overlap_style") or "dim"
+    self.page_gap.height = config:readSetting("kopt_page_gap") or 8
+    self.page_gap_width = config:readSetting("kopt_page_gap") or 8
```
Comment (yparitcher) [inline frontend/apps/reader/modules/readerview.lua]: page gap is already defined here:
https://github.com/koreader/koreader/blob/7fb2db8f02897b8ff27f529f752f9a6150ca9939/frontend/apps/reader/modules/readerview.lua#L53-L59

so you might want to change this to:
```diff
-   self.page_gap.height = config:readSetting("kopt_page_gap") or 8
+   self.page_gap.height = config:readSetting("kopt_page_gap") or G_reader_settings:readSetting("page_gap_width") or 8
```
and not define it on line 55.

also there it uses `scaleBySize` instead of a direct number.

**Comment by mustafa-001** [comment]
Aww, when trying to separate landscape and portrait page gap I just realized they are both calculated from `ReaderView.page_gap.height`, `page_gap.width` works in horizontal scrolling (left-right, which I think not much useful and currently not implemented. Maybe with few and very long lines?). 

So I'm removing width, moving height to :onreadSettings(), leaving page_gap.color only.
https://github.com/koreader/koreader/blob/7fb2db8f02897b8ff27f529f752f9a6150ca9939/frontend/apps/reader/modules/readerview.lua#L52-L60

File: frontend/apps/reader/modules/readerview.lua [hunk 2]
```diff
@@ -782,8 +778,7 @@ function ReaderView:onReadSettings(config)
     self.page_scroll = page_scroll == 1 and true or false
     self.highlight.saved = config:readSetting("highlight") or {}
     self.page_overlap_style = config:readSetting("page_overlap_style") or G_reader_settings:readSetting("page_overlap_style") or "dim"
-    self.page_gap.height = config:readSetting("kopt_page_gap") or 8
-    self.page_gap_width = config:readSetting("kopt_page_gap") or 8
+    self.page_gap.height = Screen:scaleBySize(config:readSetting("kopt_page_gap_height") or 8)
```
Comment (yparitcher) [inline frontend/apps/reader/modules/readerview.lua]: you might want
```diff
-   self.page_gap.height = Screen:scaleBySize(config:readSetting("kopt_page_gap_height") or 8)
+   self.page_gap.height = Screen:scaleBySize(config:readSetting("kopt_page_gap_height") or G_reader_settings:readSetting("page_gap_width") or 8)
```

File: frontend/apps/reader/modules/readerview.lua [hunk 3]
```diff
@@ -51,10 +51,6 @@ local ReaderView = OverlapGroup:extend{
     scroll_mode = "vertical",
     -- properties of the gap drawn between each page in scroll mode:
     page_gap = {
-        -- width in pixels (when scrolling horizontally)
-        width = Screen:scaleBySize(G_reader_settings:readSetting("page_gap_width") or 8),
```
Comment (yparitcher) [inline frontend/apps/reader/modules/readerview.lua]: if you are getting rid of this you should also clean up
https://github.com/koreader/koreader/blob/7fb2db8f02897b8ff27f529f752f9a6150ca9939/frontend/apps/reader/modules/readerview.lua#L392-L396
and
https://github.com/koreader/koreader/blob/7fb2db8f02897b8ff27f529f752f9a6150ca9939/frontend/apps/reader/modules/readerview.lua#L51

does anyone know what `scroll_mode` is used for and if it is ever set to `horizontal`?

Comment (Frenzie) [reply frontend/apps/reader/modules/readerview.lua]: I renamed scroll mode to continuous in the GUI.

File: frontend/apps/reader/modules/readerview.lua (hunk 2 above)
Comment (yparitcher) [reply frontend/apps/reader/modules/readerview.lua]: that should be 
`self.page_gap.height = Screen:scaleBySize(config:readSetting("kopt_page_gap_height") or G_reader_settings:readSetting("kopt_page_gap_hight") or 8)`

also this does not work with defaults (holding the option) with or without this change

File: frontend/apps/reader/modules/readerview.lua (hunk 3 above)
Comment (yparitcher) [reply frontend/apps/reader/modules/readerview.lua]: @Frenzie 
`scroll_mode` or `page_scroll`?

File: frontend/apps/reader/modules/readerview.lua (hunk 2 above)
Comment (yparitcher) [reply frontend/apps/reader/modules/readerview.lua]: sorry typo caused that.
to get defaults to work you need this
```diff
-    self.page_gap.height = Screen:scaleBySize(config:readSetting("kopt_page_gap_height") or 8)
+    self.page_gap.height = Screen:scaleBySize(config:readSetting("kopt_page_gap_height") or G_reader_settings:readSetting("kopt_page_gap_height") or 8)
```

File: frontend/apps/reader/modules/readerview.lua (hunk 3 above)
Comment (mustafa-001) [reply frontend/apps/reader/modules/readerview.lua]: `page_scroll` is a boolean that holds `True` in continuous mode, `False` in page mode. `scroll_mode` set `vertical` in `ReaderView` creation and doesn't change.

Comment (Frenzie) [reply frontend/apps/reader/modules/readerview.lua]: That's part of the same thing, perhaps separated specifically for that vertical vs horizontal thing. I don't know the specific plan behind 63e5e7e577a0fb18b27f0b50400e38ef650b3a71

File: frontend/apps/reader/modules/readerview.lua [hunk 4]
```diff
@@ -778,7 +773,8 @@ function ReaderView:onReadSettings(config)
     self.page_scroll = page_scroll == 1 and true or false
     self.highlight.saved = config:readSetting("highlight") or {}
     self.page_overlap_style = config:readSetting("page_overlap_style") or G_reader_settings:readSetting("page_overlap_style") or "dim"
-    self.page_gap.height = Screen:scaleBySize(config:readSetting("kopt_page_gap_height") or 8)
+    self.page_gap.height = Screen:scaleBySize(config:readSetting("kopt_page_gap_height") or
+    G_reader_settings("kopt_page_gap_height") or 8)
```
Comment (yparitcher) [inline frontend/apps/reader/modules/readerview.lua]: ``diff
-    G_reader_settings("kopt_page_gap_height") or 8)
+    G_reader_settings:readSetting("kopt_page_gap_height") or 8)
```

**Comment by yparitcher** [comment]
lgtm besides for the typo

**Comment by poire-z** [comment]
> Yeah, IIRC, kopt settings can force repaints without any specific event because... reasons? :D.

https://github.com/koreader/koreader/blob/c3a0bd4def2243b66cacf27222b62d9441ebbaff/frontend/ui/widget/configdialog.lua#L909-L913
Part of #4793, with some long discussion trying to understand how that works, that I'd rather not re-read :)
I guess that some options are only used by the kopt code, and there's no (or no need for) Lua code to handle some event, as there's no real action - except having kopt do its stuff with the updated settings.
For this one (page_gap.height), that's our Lua code that draws the gray line, so it looks allright to have an event and an handler (or we could just re-read the setting each time we draw a page, but no need to bother changing what's here I guess).

File: frontend/apps/reader/modules/readerview.lua [hunk 5]
```diff
@@ -782,6 +773,8 @@ function ReaderView:onReadSettings(config)
     self.page_scroll = page_scroll == 1 and true or false
     self.highlight.saved = config:readSetting("highlight") or {}
     self.page_overlap_style = config:readSetting("page_overlap_style") or G_reader_settings:readSetting("page_overlap_style") or "dim"
+    self.page_gap.height = Screen:scaleBySize(config:readSetting("kopt_page_gap_height") or
+    G_reader_settings:readSetting("kopt_page_gap_height") or 8)
```
Comment (poire-z) [inline frontend/apps/reader/modules/readerview.lua]: (nitpick: I'd like a bit more indentation on that 2nd line, to avoid confusion and make it clear it's continued from previous)

**Comment by Frenzie** [comment]
___
*[frontend/ui/data/koptoptions.lua, line 88 at r6](https://reviewable.io/reviews/koreader/koreader/5705#-LxW_IaV8mfK-Iglr0KW:-LxW_IaW8ad5luR-7gbH:b-zc79rk) ([raw file](https://github.com/koreader/koreader/blob/455f1cdd67dd830d51b41c2dc29716d244e658b1/frontend/ui/data/koptoptions.lua#L88)):*
> ```Lua
>             {
>                 name = "page_gap_height",
>                 name_text = S.PAGE_GAP,
> ```

I'd just use the string here instead of adding it to that strings.lua file. For some reason those strings were globals back in 2013 but separating them from their surroundings just makes translation harder.

**Comment by poire-z** [comment]
Is this ready to merge?
(I think @Frenzie last comment about `S.PAGE_GAP` should be adressed in a dedicated "chore" PR, where we'd do all of them in one batch.)

**Comment by Frenzie** [comment]
If you have no further comments, sure. I haven't tested it.

**Comment by poire-z** [comment]
> I haven't tested it.

Neither did I, but that's what users are for :)

I guess the Circle CI build (repeatedly) fails because of the transiflex>weblate move, and this PR base commit is not adapted to the CircleCI updated tests?

**Comment by Frenzie** [comment]
Sort of, a minor accident on my part caused the submodule reference to become invalid.

My plan was simple, to update Weblate with the latest from Transifex when the time came, the end.

I did that, but it turned out Weblate had already attracted a few dozen new translations. So instead of overwriting I had to redo it by merging the two translations, and I semi-accidentally rebased that one commit. I wasn't fully cognizant of the fact that it'd change the commit hash if I just kept it in there.

tl;dr Now older commits referencing the submodule as it was can't find their reference.
